capsules/extra/src/nonvolatile_storage_driver.rs

```diff
@@ -38,6 +79,45 @@
 //! +-----------------------------------------------------------------+
 //! ```
 //!
+//! Example nonvolatile storage layout:
+//!
+//! ```text
+//!     ╒════════ ← Start of kernel region
+//!     │
+//!     │
+//!     │
+//!     ╞════════ ← Start of userspace region
+//!     ├──────── ← Start of App 1's region header
+//!     │ App 1's magic value (usize) = HEADER_V1 ^ ShortID ^ length
+//!     │ App 1's ShortID (u32)  
+//!     │ region 1 length (usize)
```

Comment:
With the disclaimer that this is pretty in the weeds given the context of this PR and how this capsule will be used, but:

This seems like an undesirable format because it requires speculatively checking for each version, and a chance of uncertainty. Or it requires each future version to guard against collisions with previous versions.

Imagine there was a version 2 and that the logic starts by assuming the region has a v2 header. What we wouldn't want is for a version 1 header to be decoded as a version 2. But it's not clear to me (given the general structure here) how a v2 header would be constructed such that it is impossible (particularly since we don't know anything about ShortId other than it is not 0) that these three fields and whatever data the app stored in its region could never match a valid v2 header.

That's why its common for the version to be stored separately and then there is no guessing what version the frame is.

This could be 

```
- 8 bits  - 24 bits                               -
| version | length                                |
| shortid                                         |
| xor                                             |
```

to keep it the same size.

Reply:
That makes sense to have the version be stored separately instead of having to compute it and check it against all previous versions.

I just want to clarify the frame layout you showed. 

The first 32 bits are a concatenation of a u8 version and a 24 bit length. Then the next 32 bits are the shortid. And the last 32 bits are an xor between the last 2 32 bits.

Reply:
Yes, you are reading that diagram correctly

Reply:
I went ahead and implemented this new header format.

However, I'm running into issues on determining if headers are valid. My first idea was to check if 
`xor == (version | length) ^ shortid`

I quickly realized this wouldn't work since I ran into an area of flash where all three fields of this frame were read in as 0. This satisfied the validity check since `0 == 0^0`.

Should I go back to matching the version number against all previous versions to show validity (in combination with my xor check)? Or is there another way I'm not thinking of checking validity with this format?

Reply:
Don't use version 0, start at 1 :)

[also be sure to shift `version << 24` in actual code if you aren't; just read this comment]

Reply:
If it's useful, you could split the current 8-bit version into 4-bit "magic" + 4-bit "version", such that there's something slightly more unique; not sure if that's necessary though.

Reply:
I think the version number should be unique on its own without an extra magic number. I was using 1 as the V1 value as well.

I think I misinterpreted what Brad said [here](https://github.com/tock/tock/pull/4109#discussion_r1732888856) and thought that he didn't want me to check if the version number itself is one of the valid version numbers. But I think he meant that I shouldn't try to calculate the version from the result of an XOR operation (it's clearer just to read it in plainly).

So I think I'll check equality of the version I read as well as check that the `xor` field matches up.

Reply:
The implementation of this change is up in this commit https://github.com/tock/tock/pull/4109/commits/327743ec1958a9558fccc00213aba778d336e801